src/i_winmusic.c

```diff
@@ -621,12 +638,390 @@ static void SendSysExMsg(int time, const byte *data, int length)
         // GS (SC-88 or higher). Preserve the composer's intent.
         MIDI_ResetFallback();
         use_fallback = false;
+
+        // Use default device for EMIDI.
+        emidi_device = EMIDI_DEVICE_GENERAL_MIDI;
+    }
+}
+
+static void SendProgramMsg(int time, int channel, int program,
+                           midi_fallback_t *fallback)
+{
+    switch ((int)fallback->type)
+    {
+        case FALLBACK_BANK_MSB:
+            SendShortMsg(time, MIDI_EVENT_CONTROLLER, channel,
+                         MIDI_CONTROLLER_BANK_SELECT_MSB, fallback->value);
+            SendShortMsg(0, MIDI_EVENT_PROGRAM_CHANGE, channel, program, 0);
+            break;
+
+        case FALLBACK_DRUMS:
+            SendShortMsg(time, MIDI_EVENT_PROGRAM_CHANGE, channel,
+                         fallback->value, 0);
+            break;
+
+        default:
+            SendShortMsg(time, MIDI_EVENT_PROGRAM_CHANGE, channel, program, 0);
+            break;
+    }
+}
+
+static void SetLoopPoint(void)
+{
+    unsigned int i;
+
+    for (i = 0; i < song.num_tracks; ++i)
+    {
+        MIDI_SetLoopPoint(song.tracks[i].iter);
+        song.tracks[i].saved_end_of_track = song.tracks[i].end_of_track;
+        song.tracks[i].saved_elapsed_time = song.tracks[i].elapsed_time;
+    }
+    song.saved_elapsed_time = song.elapsed_time;
+}
+
+static void CheckFFLoop(midi_event_t *event)
+{
+    if (event->data.meta.length == sizeof(ff_loopStart) &&
+        !memcmp(event->data.meta.data, ff_loopStart, sizeof(ff_loopStart)))
+    {
+        SetLoopPoint();
+        song.ff_loop = true;
+    }
+    else if (song.ff_loop && event->data.meta.length == sizeof(ff_loopEnd) &&
+             !memcmp(event->data.meta.data, ff_loopEnd, sizeof(ff_loopEnd)))
+    {
+        song.ff_restart = true;
     }
 }
 
+static boolean AddToBuffer(unsigned int delta_time, midi_event_t *event,
```

Comment:
> Maybe I can generalize this for FluidSynth at least, but that's for the future.

I think I should take this entire function and have it at the `midifile.h` level. That would help the FluidSynth driver in the future if it's ever made more low level, it could also help OPL (for custom loops if nothing else), and I just had an idea of writing a VST driver where the user simply drops their VST in a Woof/VST folder (e.g `syxg50.dll` or `SOUND Canvas VA.dll`) and it is automatically read by Woof. No more loopMIDI, FSMP, SAVIHost, or MIDIMapper needed.

Reply:
> I think I should take this entire function and have it at the `midifile.c/.h` level. That would help the FluidSynth driver in the future if it's ever made more low level, it could also help OPL.

This is basically PrBoom+ approach. I want too keep FluidSynth's sequencer and use [fluid_sequencer_send_at()](https://www.fluidsynth.org/api/group__sequencer.html#ga83314f4ad773979841afe50cc2efd83f) function.

> I just had an idea of writing a VST driver where the user simply drops their VST in a Woof/VST folder (e.g `syxg50.dll` or `SOUND Canvas VA.dll`) and it is automatically read by Woof.

I think this is too ambitious, the VST API is not that simple. Choco/Crispy will definitely not accept this. Also the user still needs to get the VST plugins, so this won't be a complete solution.

Reply:
I think the [licensing issues](https://steinbergmedia.github.io/vst3_dev_portal/pages/FAQ/Licensing.html#q-i-am-a-developercompany-and-i-want-to-develop-and-distribute-a-vst-2-plug-in-andor-host-in-binary-form) with VST kills the idea anyway (can't distribute a VST2 host anymore).

> Q: I am a developer/company and I want to develop and distribute a VST 2 plug-in and/or host in binary form.
If you have signed the VST 2 license agreement (before October 2018), you can.
If not, you are not allowed to distribute it!

Reply:
> I want too keep FluidSynth's sequencer and use [fluid_sequencer_send_at()](https://www.fluidsynth.org/api/group__sequencer.html#ga83314f4ad773979841afe50cc2efd83f) function.

Do you think it's worth making major changes to `i_flmusic.c` in order to support the following niche features?
1. Partial SysEx support
2. Custom loop points

For (1), looking at the FluidSynth source code, this is only GM/GS/XG reset (not needed because you can just restart the FluidSynth synth after each song) and for GS songs that change drum channel using SysEx (doesn't exist in any Doom wad, as far as I know). Also, GS/XG support depends on the soundfont and almost all of them only have GM with a few GS banks filled out. FluidSynth already defaults to GS and then falls back when needed (e.g. MAYhem19 title music). So no fallback support is needed either.

Reply:
Yes, this is only for loop points and EMIDI. Of course, this is a low priority, only after `i_winmusic.c` has been accepted into Choco/Crispy. Currently, EMIDI/loop points will be Windows-exclusive, but that's feels wrong, so I'll at least give it a try.